ls1intum/ArTEMiS
# Complaints for modeling exercises

### Checklist

### Description
In order to give students the opportunity to complain about assessments we introduced a complaint mechanism that is already used in text exercises (currently disabled). We want to provide the complaint feature also in modeling exercises now. 

Requirements:
- Students have the possibility to submit a complaint about a modeling assessment in the assessment result view 
- Tutors see the list of complaints in the tutor exercise dashboard
- Tutors do not see complaints about their own assessments
- Tutors can open a complaint via the tutor dashboard to review it
- In the review complaint view tutors see the submission with its original assessment
- They can submit a reply and update the assessment (both together with one button "Update assessment"), if the complaint is accepted 
- They can send a reply and reject the complaint with a second button "Reject complaint"
- Tutors cannot respond (accept or reject) to complaints of one of their own assessments
- Updating the assessment is handled in the server as follows:
  - The original result is stored in the _resultBeforeComplaint_ field of a complaint as a JSON string for future lookup
  - In the result of the corresponding submission the _score_ and _feedback list_ is updated, the _assessor_ is not updated
  - The assessor of the updated assessment is stored in the _reviewer_ field of the complaint response
- Students can only have up to 3 complaints at the same time that are either rejected or open/unhandled (accepted complaints are not taken into account for the max. limit)
- Students can only create a complaint up to one week after the result _completionDate_

### Steps for Testing
1. Log in to ArTEMiS as a student
2. Navigate to assessed modeling submission
3. Click on "Complain"
4. Enter and submit a complaint
5. Log in to ArTEMiS as a tutor (not the assessor of the submission from above)
6. Navigate to tutor exercise dashboard for the corresponding exercise
7. See list of complaints and click on "Evaluate the complaint"
8. Reply to complaint
9. Log in as student again
10. Check response for previously submitted complaint

### Screenshots
Submit a complaint
![Submit Complaint](https://user-images.githubusercontent.com/5919138/57194825-9ab08200-6f4b-11e9-9bd3-83986b3ecf65.JPG)

Evaluate complaint as tutor
![Evaluate Complaint](https://user-images.githubusercontent.com/5919138/57194829-a8fe9e00-6f4b-11e9-9188-22f695af4689.JPG)

Complaint accepted - tutor view
![Complaint accepted](https://user-images.githubusercontent.com/5919138/57194831-b2880600-6f4b-11e9-80e6-41e48ac74573.JPG)

Complaint rejected - student view
![Complaint rejected - student view](https://user-images.githubusercontent.com/5919138/57194833-bae04100-6f4b-11e9-89ea-dd82962eb5a1.JPG)

Complaint currently not allowed
![Complaint currently not allowed](https://user-images.githubusercontent.com/5919138/57194836-c29fe580-6f4b-11e9-8053-d9fd7e93defe.JPG)
